From 3a003733e6c8081c07fec51a77f65d3d5d8a1094 Mon Sep 17 00:00:00 2001
From: Nick Mathewson <nickm@torproject.org>
Date: Wed, 21 Apr 2010 12:25:29 -0400
Subject: [PATCH] Detect broken unsetenv at unit-test runtime

If we have an unsetenv function that doesn't work, we can't run the
main/base_environ unit test, so we should skip it.
---
 test/regress.c |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/test/regress.c b/test/regress.c
index 513c3c8..cda0892 100644
--- a/test/regress.c
+++ b/test/regress.c
@@ -1830,6 +1830,18 @@ test_base_environ(void *arg)
 	int i, n_methods=0;
 	const char *defaultname;
 
+	/* See if unsetenv works before we rely on it. */
+	setenv("EVENT_NOWAFFLES", "1", 1);
+	unsetenv("EVENT_NOWAFFLES");
+	if (getenv("EVENT_NOWAFFLES") != NULL) {
+#ifndef _EVENT_HAVE_UNSETENV
+		TT_DECLARE("NOTE", ("Can't fake unsetenv; skipping test"));
+#else
+		TT_DECLARE("NOTE", ("unsetenv doesn't work; skipping test"));
+#endif
+		tt_skip();
+	}
+
 	basenames = event_get_supported_methods();
 	for (i = 0; basenames[i]; ++i) {
 		methodname_to_envvar(basenames[i], varbuf, sizeof(varbuf));
-- 
1.6.6.1

